src/context/mod.rs

```diff
@@ -849,6 +854,351 @@ bitflags! {
+    for InfoContextFieldBottomLevelData
+{
+    fn from(value: BTreeMap<String, T>) -> Self {
+        Self(
+            value
+                .into_iter()
+                .map(|e| (e.0, e.1.into()).into())
+                .collect(),
+        )
+    }
+}
+
+impl<T: Into<InfoContextBuilderFieldBottomLevelValue>> From<HashMap<String, T>>
+    for InfoContextFieldBottomLevelData
+{
+    fn from(value: HashMap<String, T>) -> Self {
+        Self(
+            value
+                .into_iter()
+                .map(|e| (e.0, e.1.into()).into())
+                .collect(),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct InfoContextBuilder<'a> {
+    context: &'a InfoContext,
+    sections: InfoContextTreeData,
+}
+impl<'a> InfoContextBuilder<'a> {
+    fn add_bottom_level_field(
+        &self,
+        key: &str,
+        value: &InfoContextBuilderFieldBottomLevelValue,
+    ) -> RedisResult<()> {
+        use InfoContextBuilderFieldBottomLevelValue as BottomLevel;
+
+        match value {
+            BottomLevel::String(string) => add_info_field_str(self.context.ctx, key, string),
+            BottomLevel::I64(number) => add_info_field_long_long(self.context.ctx, key, *number),
+            BottomLevel::U64(number) => {
+                add_info_field_unsigned_long_long(self.context.ctx, key, *number)
+            }
+            BottomLevel::F64(number) => add_info_field_double(self.context.ctx, key, *number),
+        }
+        .into()
+    }
+    /// Adds fields. Make sure that the corresponding section/dictionary
+    /// have been added before calling this method.
+    fn add_top_level_fields(&self, fields: &InfoContextFieldTopLevelData) -> RedisResult<()> {
+        use InfoContextBuilderFieldTopLevelValue as TopLevel;
+
+        fields.iter().try_for_each(|(key, value)| match value {
+            TopLevel::Value(bottom_level) => self.add_bottom_level_field(key, bottom_level),
+            TopLevel::Dictionary { name, fields } => {
+                std::convert::Into::<RedisResult<()>>::into(add_info_begin_dict_field(
+                    self.context.ctx,
+                    name,
+                ))?;
+                fields
+                    .iter()
+                    .try_for_each(|f| self.add_bottom_level_field(&f.0 .0, &f.0 .1))?;
+                add_info_end_dict_field(self.context.ctx).into()
+            }
+        })
+    }
+
+    fn finalise_data(&self) -> RedisResult<()> {
+        self.sections
+            .iter()
+            .try_for_each(|(section_name, section_fields)| -> RedisResult<()> {
+                Into::<RedisResult<()>>::into(add_info_section(
+                    self.context.ctx,
+                    Some(section_name),
+                ))?;
+
+                self.add_top_level_fields(section_fields)
+            })
+    }
+
+    /// Sends the info accumulated so far to the [`InfoContext`].
+    pub fn build_info(self) -> RedisResult<&'a InfoContext> {
```

Comment:
If I understand correctly, this will return an error if Redis returns status REDISMODULE_ERR. If so will lose the InfoContext and will not be able to continue add more sections. I believe this is not the expected behaviour, if Redis returns REDISMODULE_ERR it means that the section was not requested and we should not add it, but it does not mean we should stop adding more sections.

Reply:
Please correct me if I misunderstood the code ...